rewrite: Bitbucket server, and minor tweaks to BROWSER

derimagia opened the pull request:
Added bitbucket server support. I didn't add functionality for the prefix yet, i'd rather do that for all of them at once (Having a way to supply replacements for the "server" with the root path would take care of Custom Schemes (http vs https), Gitlab having a separate ssh vs http domain, and having custom roots). There any objections to this? Having separate options for every service sounds a little strange rather than supporting them everywhere

From the tests, `gitopen.gitlab.port` should no longer be needed since it just removes the protocol anyway i believe.

This also makes it so you can supply "BROWSER" on all platforms, or was there a reason why this wasn't allowed? Made testing slightly easier to allow it. I had to hardcode "echo" to not be exported to /dev/null since before it was like this for tests, if we add options using getopts a verbosity option would be nice here

derimagia commented:
I also implemented #80 here, it looks like https://bitbucket.org/guyzmo/git-repo/src?at=bugfix/conftest_fix worked without the ref (or HEAD) at all

paulirish commented:
@derimagia take look at these two results:

* https://bitbucket.org/guyzmo/git-repo/src/HEAD?at=bugfix/conftest_fix
* https://bitbucket.org/guyzmo/git-repo/src/ef1e8c9f87c82f22675107505f73b307025744b7/?at=bugfix%2Fconftest_fix

you can see a few changes in the second. (commits from `2016-12-28`).  Thats what leads me to believe there's not an easy fix for the "slash branches"

paulirish commented:
@derimagia  I like the look of the refactors in here. Very nice!
Also I'm going to rebase this branch against master. Will let you know when that's complete.

> Having separate options for every service sounds a little strange rather than supporting them everywhere

+1 to this.

> gitopen.gitlab.port should no longer be needed 

yeah i'm hoping that's the case, it seemed like a stretch anyway.

> This also makes it so you can supply "BROWSER" on all platforms, or was there a reason why this wasn't allowed?

no reason. i like having it available everywhere. this is much better.

--------

I'll hold off on merging until we hear confirmation from @nwinkler about the details of the bb server stuff.   
@nwinkler, FYI the tests in this PR are a little different than the test in #83 so we'll need your help to figure out what the I/O is.

paulirish commented:
@derimagia fwiw you have some `abc` file/submodule in this PR. Looks like I'll clean that up and publish a new `rewrite` branch with this commit integrated.

paulirish commented:
@derimagia so far it looks like the `BROWSER=echo` line in `setup` isn't making its way to inside of the @test blocks on Mac OS. :/

Changing it to `  export BROWSER=echo` actually seems to be safe. Exports from inside of bats appear to not affect the real shell, so i believe this is safe.  wdyt?

derimagia commented:
Interesting, https://github.com/paulirish/git-open/blob/fbc508a940c3ad25117807a87e5d96e4da98c220/git-open#L83 has it parse the rev for the head even when the branch is different, maybe it didn't work before or something.

I added the ref to bitbucket urls, as expected this messes with the tests if we don't have a static commit, but i'll take a look at this later tonight

I removed the directory (thought I did it, whoops) and added the export. I'm using mac too, strange It must be doing something odd, we may be able to have just lifted it up but I'll just export it for now.

paulirish commented:
@derimagia btw take a look at `rewrite`: https://github.com/paulirish/git-open/compare/rewrite

paulirish commented:
@derimagia can you work from the `rewrite` branch? looks like the `rev` adjustments you made i don't have there. wanna add a commit for that?  I'll open a PR just to track the overall thing.

derimagia commented:
Changed the branch, it still has `refute_output --partial "//"` in it - I assumed you put there as a marker?

paulirish commented:
perfect. this is great.
